jwst/pipeline/calwebb_image3.py

```diff
@@ -69,22 +67,23 @@ def process(self, input_data):
 
         self.source_catalog.save_results = self.save_results
 
-        with datamodels.open(input_data, asn_exptypes=asn_exptypes) as input_models:
+        # Only load science members from input ASN;
+        # background and target-acq members are not needed.
+        input_models = self._load_input_as_library(input_data)
+
+        if (self.output_file is None) and hasattr(input_models.asn["products"][0], "name"):
             # If input is an association, set the output to the product name.
-            if self.output_file is None:
-                try:
-                    self.output_file = input_models.meta.asn_table.products[0].name
-                except (AttributeError, IndexError):
-                    pass
+            self.output_file = input_models.asn["products"][0]["name"]
 
-            # Check if input is single or multiple exposures
-            try:
-                has_groups = len(input_models.group_names) >= 1
-            except (AttributeError, TypeError, KeyError):
-                has_groups = False
+        # Check if input is single or multiple exposures
+        has_groups = len(input_models) > 1
 
-            if isinstance(input_models, ModelContainer) and has_groups:
-                if is_moving_target(input_models):
+        with input_models:
+            if has_groups:
+                model = input_models.borrow(0)
+                is_moving = is_moving_target(model)
+                input_models.shelve(model, 0, modify=False)
```

Comment:
```suggestion
                with input_models:
                    model = input_models.borrow(0)
                    is_moving = is_moving_target(model)
                    input_models.shelve(model, 0, modify=False)
```
The current code passes an open library to `assign_mtwcs`, `tweakreg` and `skymatch`. With this suggestion (and some change to non-modified lines which I can't add a suggestion for) the library can be closed before passing it to the steps.